pkg/odo/cli/dev/dev.go

```diff
@@ -193,29 +234,64 @@ func (o *DevHandler) RegenerateAdapterAndPush(pushParams common.PushParameters,
 }
 
 func regenerateComponentAdapterFromWatchParams(parameters watch.WatchParameters) (common.ComponentAdapter, error) {
-
-	// Parse devfile and validate. Path is hard coded because odo expects devfile.yaml to be present in the pwd/cwd.
-
 	devObj, err := ododevfile.ParseAndValidateFromFile(location.DevfileLocation(""))
 	if err != nil {
 		return nil, err
 	}
 
+	changed, err := haveEndpointsChanged(parameters.InitialDevfileObj, devObj)
+	if err != nil {
+		return nil, err
+	} else if changed {
+		fmt.Printf("\ntotal number of endpoints in the devfile have changed; please run `odo dev` again\n\n")
+	}
+
 	platformContext := kubernetes.KubernetesContext{
 		Namespace: parameters.EnvSpecificInfo.GetNamespace(),
 	}
 
 	return adapters.NewComponentAdapter(parameters.ComponentName, parameters.Path, parameters.ApplicationName, devObj, platformContext)
+}
+
+func haveEndpointsChanged(oldDevfile, newDevfile parser.DevfileObj) (bool, error) {
+	oldEndpoints, err := libdevfile.GetAllEndpointsFromDevfile(oldDevfile)
+	if err != nil {
+		return false, err
+	}
+
+	newEndpoints, err := libdevfile.GetAllEndpointsFromDevfile(newDevfile)
+	if err != nil {
+		return false, nil
+	}
 
+	if !reflect.DeepEqual(oldEndpoints, newEndpoints) {
+		return true, nil
+	}
+	return false, nil
+}
+
+func printPortForwardingInfo(portPairs map[string][]string, out io.Writer) {
+	portForwardURLs := ""
```

Comment:
This is a good candidate to use `strings.Builder`: https://pkg.go.dev/strings#Builder